### PR TITLE
Feature/access level

### PR DESCRIPTION
**Pull request by joe94**
This PR includes: 
- The access level feature which allows requests to have elevated permissions based on a JWT claim containing some capability.
- A fix to the behavior around an item's TTL.

### CODE REVIEW

File: store/transport.go [hunk 1]
```diff
@@ -49,19 +55,10 @@ var (
...
-type requestConfig struct {
-	Validation    validationConfig
-	Authorization authorizationConfig
-}
-
-type validationConfig struct {
-	MaxTTL time.Duration
+type transportConfig struct {
+	AccessLevelAttributeKey string
+	ItemMaxTTL              time.Duration
```
Comment (kristinaspring) [inline store/transport.go:60]: so flat! 💯

File: auth/accessLevel.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	if len(config.CapabilitySource.Path) < 1 {
+		config.CapabilitySource.Path = defaultAccessLevelPath
+	}
+}
+
+func newAccessLevel(config *accessLevelConfig) accessLevel {
+	validateAccessLevelConfig(config)
+
+	resolver := func(attributes bascule.Attributes) int {
+		capabilitiesClaim, ok := bascule.GetNestedAttribute(attributes, config.CapabilitySource.Path...)
```
Comment (kristinaspring) [inline auth/accessLevel.go:77]: To keep this generic, I don't think we should call them capabilities - someone could match against a differently named list if they wanted to.  I'm not against leaving it though because my pick would be something like `attributes` which is pretty generic, and becomes confusing when returning `Access Level Attribute Value`s.

TLDR: idk, you probably shouldn't change anything.

File: auth/accessLevel.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	if len(config.CapabilitySource.Name) < 1 {
+		config.CapabilitySource.Name = defaultAccessLevelCapabilityName
+	}
+
+	if len(config.CapabilitySource.Path) < 1 {
+		config.CapabilitySource.Path = defaultAccessLevelPath
+	}
+}
+
+func newAccessLevel(config *accessLevelConfig) accessLevel {
```
Comment (kristinaspring) [inline auth/accessLevel.go:73]: `newAccessLevel()` feels really generic...since we're creating an accessLevel that checks that the attributes found at the key given contain the name given, my best suggestion is `newContainsAccessLevel()`

File: auth/bearerTokenFactory.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	"emperror.dev/emperror"
+	"github.com/dgrijalva/jwt-go"
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/bascule/basculehttp"
+	"github.com/xmidt-org/bascule/key"
+)
+
+const jwtPrincipalKey = "sub"
+
+// accessLevelBearerTokenFactory extends basculehttp.BearerTokenFactory by letting the user of the factory inject an access level attribute to the jwt token. Application code
```
Comment (kristinaspring) [inline auth/bearerTokenFactory.go:17]: this is...............too long. 🙁 Can you put it on a few lines?

File: auth/bearerTokenFactory.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+// should handle case in which the value is not injected (i.e. basic auth tokens)
```
Comment (kristinaspring) [inline auth/bearerTokenFactory.go:18]:
```suggestion
// should handle case in which the value is not injected (i.e. basic auth tokens).
```

**Review by kristinaspring — CHANGES_REQUESTED** [review]
Mostly comments on names of things.  The whole flow and organization looks great! 👍

File: auth/constructor.go [hunk 6]
```diff
@@ -47,11 +49,12 @@ func providePrimaryBasculeConstructorOptions(apiBase string) fx.Option {
 					return nil
 				}
 				in.Logger.Log(level.Key(), level.DebugValue(), xlog.MessageKey(), "building bearer token factory option", "server", "primary")
-				return basculehttp.WithTokenFactory("Bearer", basculehttp.BearerTokenFactory{
-					DefaultKeyId: in.DefaultKeyID,
+				return basculehttp.WithTokenFactory("Bearer", accessLevelBearerTokenFactory{
+					DefaultKeyID: in.DefaultKeyID,
```
[the remark this answers is not in the document]
Comment (joe94) [reply auth/constructor.go:53]: 😂  the change we've all been waiting for

File: auth/provide.go [hunk 7]
```diff
@@ -42,6 +42,17 @@ func ProvidePrimaryServerChain(apiBase string) fx.Option {
 		))
 }
 
+// accessLevelAttributeKeyAnnotated shares the access level attribute key which outside packages such as argus/store can
+// as parameters.
+func accessLevelAttributeKeyAnnotated() fx.Annotated {
+	return fx.Annotated{
+		Name: "access_level_attribute_key",
```
[the remark this answers is not in the document]
Comment (joe94) [reply auth/provide.go:49]: 💯  Both options seemed pretty odd for me at the time I added it. Does uber/fx have a standard? It seemed fine to me having `argus/auth` emitting the value since it originates there 🤷‍♂️

File: store/provide.go [hunk 8]
```diff
@@ -48,22 +50,24 @@ type StoreOut struct {
...
+// NewHandlers initializes all handlers that will be needed for the store endpoints.
+func NewHandlers(in StoreIn) StoreOut {
+	var itemMaxTTL time.Duration
 
-	return StoreOut{
-		SetItemHandler:     newSetItemHandler(cfg, in.Store),
-		GetItemHandler:     newGetItemHandler(cfg, in.Store),
-		GetAllItemsHandler: newGetAllItemsHandler(cfg, in.Store),
-		DeleteKeyHandler:   newDeleteItemHandler(cfg, in.Store),
+	in.Unmarshaler.UnmarshalKey("itemMaxTTL", &itemMaxTTL)
```
[the remark this answers is not in the document]
Comment (joe94) [reply store/provide.go:57]: Good suggestion! I'd rather focus on that in a separate PR though https://github.com/xmidt-org/argus/issues/113

File: auth/accessLevel.go (hunk 2 above)
Comment (joe94) [reply auth/accessLevel.go:77]: We can revisit this later if it becomes an issue 👍

**Review by kristinaspring — APPROVED** [review]
awesome 👍

File: auth/provide.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kristinaspring) [reply auth/provide.go:49]: yeah I'm good with that 👍 this is like one of the artifacts it's providing.

**Review by kristinaspring — APPROVED** [review]

File: deploy/packaging/argus_spruce.yaml [hunk 9]
```diff
@@ -124,23 +124,12 @@ dynamo:
   secretKey: (( grab $AWS_ACCESS_SECRET || "secretKey" ))
 
 
+itemMaxTTL: (( grab $ITEM_MAX_TTL || "24h"
```
Comment (kcajmagic) [inline deploy/packaging/argus_spruce.yaml:127]:
```suggestion
itemMaxTTL: (( grab $ITEM_MAX_TTL || "24h" ))
```

**Review by kcajmagic — CHANGES_REQUESTED** [review]
Looks good, one minor change

**Review by kcajmagic — APPROVED** [review]
